File: rust/src/errors/broker.rs
use ::std::error::Error as StdError;

use ::thiserror::Error;

/// Error type for broker-specific errors (for example, from NATS or JetStream)
/// that provides a unified error interface for the messaging system.
#[derive(Error, Debug)]
#[error(transparent)]
pub struct BrokerError(#[from] Box<dyn StdError + Send + Sync>);

impl BrokerError {
  /// Creates a new BrokerError from any error type that implements StdError.
  pub fn new<E>(err: E) -> Self
  where
    E: StdError + Send + Sync + 'static,
  {
    Self(Box::new(err))
  }
}

File: go/errors/broker.go
package errors

// BrokerError represents errors from broker-specific operations
// (for example, from NATS or JetStream) that provides a unified error
// interface for the messaging system.
type BrokerError struct {
	err error
}

// Error implements the error interface for BrokerError.
func (e *BrokerError) Error() string {
	return e.err.Error()
}

// Unwrap returns the underlying error for error chain inspection.
func (e *BrokerError) Unwrap() error {
	return e.err
}

// NewBrokerError creates a new BrokerError from any error type.
func NewBrokerError(err error) *BrokerError {
	return &BrokerError{err: err}
}
